src/cargo/ops/common_for_install_and_uninstall.rs

```diff
@@ -521,6 +524,54 @@ pub fn path_source(source_id: SourceId, config: &Config) -> CargoResult<PathSour
     Ok(PathSource::new(&path, source_id, config))
 }
 
+/// A holder for a user-facing string that helps them run
+/// cargo install --git with correct arguments
+enum InstallSuggestion {
+    /// Hint to append to the error message that the user can
+    /// run to install a package from a Git Repo
+    NormalisedGitUrl(String),
+    /// Failed to generate an install suggestion.
+    None,
+}
+
+/// Returns None if the Registry Dependency was not a GitUrl, otherwise
+/// return a suggestion with the https-normalised GitUrl for user
+/// to rerun cargo install with
+fn was_git_url_miscategorised_as_a_registry_dep(dep: &Dependency) -> InstallSuggestion {
+    if dep.source_id().is_registry() {
+        if let Ok(git_url) = parse(BStr::new(dep.package_name().as_bytes())) {
+            let final_git_url: Option<InternedString> = match git_url.scheme {
+                // cargo doesn't support cargo install git@ urls, so
+                Scheme::Ssh | Scheme::Git => {
+                    if let (Some(host), Some(owner)) = (git_url.host(), git_url.user()) {
+                        let https_git_url = format!(
+                            "https://{host}/{owner}/{repo_name}",
+                            // repo_name = git_url.name()
+                            repo_name = "TODO"
```

Comment:
the reason this `repo_name = "TODO"` doesn't cause `test_install_from_https_git_generate_suggestion` to fail is the same reason why `test_install_from_git_generate_suggestion` is now failing with `error: the `--version` provided, `bitbucket.org:jcmoyer/rust-tictactoe.git`, is not a valid semver version` 

Since I opened the PR, cargo started supporting `crate@version`  command-line arguments 
https://github.com/rust-lang/cargo/commit/2806270ef63152f9ddf80ea10fad4444c2293f7c
and now the test input `git@bitbucket.org:jcmoyer/rust-tictactoe.git` is parsed as a version string (and fails with the error in the CI logs). 

the code path I added only deals with strings that weren't erroneously parsed as version strings, so the section I highlighted above won't be traversed, because `git@` strings now fail to parse as version strings before they get here.

I propose removing this chunk of code and removing the failing test to focus on supporting actionable errors for https git URLs only. 

Please let me know what you think

Reply:
actioned here
https://github.com/rust-lang/cargo/pull/10522/commits/f08e95bf0029d32979375a00d3ea5e5754b98f48